File: python-api/Mandlebrot.py
import cmath
import numpy as np
from ImageBuilder import createImageFromArrayAsRawBytes

UpperIterationLimit = 255
UpperSquaredLimit = 4
Height = 800
Width = 800

def IsInTheSet(real, imag):
    c = complex(real, imag)
    if (abs(c.real) > 2 or abs(c.imag) > 2):
                return False            

    return not(Process(c.real, c.imag) < UpperIterationLimit)

def getImageOfRange(upperLeft, lowerRight): 
    upperLeft = complex(upperLeft[0], upperLeft[1])       
    lowerRight = complex(lowerRight[0], lowerRight[1]) 
    x_resolution = (lowerRight.real - upperLeft.real) / Width
    y_resolution = (upperLeft.imag - lowerRight.imag) / Height
    depthValues = np.empty(Width * Height, dtype=int).reshape(Width,Height)
    x = 0
    for real in np.arange(upperLeft.real, lowerRight.real, x_resolution):
        y = 0
        for imag in np.arange(upperLeft.imag, lowerRight.imag, -y_resolution):        
            depthValues[x,y] = Process(real, imag)
            y = y + 1
        
        x = x + 1           

    return createImageFromArrayAsRawBytes(depthValues);       
        

def Process(real, imag):
    c = complex(real, imag)
    depth = 0
    Zn = complex(0, 0)
    while (IsInRange(Zn) and depth < UpperIterationLimit):
        Zn = Zn * Zn + c
        depth = depth + 1
    
    return depth       


def IsInRange(value):
    polarForm = cmath.polar(value)
    return (pow(polarForm[0], 2) <= UpperSquaredLimit)
        

File: rust-api/src/mandlebrot.rs
use num_complex::Complex64;

pub fn upper_iteration_limit() -> i32{
    255
}

pub fn upper_squared_limit() -> i32{
    4
}

pub fn is_in_the_set(value: Complex64) -> bool {
    if value.re.abs() > 2.0 || value.im.abs() > 2.0 {
                return false;            
    }

    return !(process(value) < upper_iteration_limit());
}

pub fn get_image_of_range(upper_left: Complex64, lower_right: Complex64) -> () {
    const HEIGHT: usize = 800;
    const WIDTH: usize = 800;

    let x_resolution = (lower_right.re - upper_left.re) / (WIDTH as f64);
    let y_resolution = (upper_left.im - lower_right.im) / (HEIGHT as f64);
    let mut depth_values = vec![vec![0; WIDTH]; HEIGHT];
    let mut x = 0;
    let mut real = upper_left.re;
    while  real < lower_right.re
    {
        let mut y = 0;
        let mut imag = lower_right.im;
        while imag < upper_left.im 
        {
            depth_values[x][y] = process(Complex64{re: real, im: imag});
            
            y = y + 1;
            if y >= HEIGHT{
                break;
            }

            imag = imag + y_resolution
        }

        x = x + 1;
        if x >= WIDTH{
            break;
        }  

        real = real + x_resolution;    
    }
    
}


pub fn process(value: Complex64) -> i32 {
    let mut depth = 0;
    let mut zn = Complex64{re:0f64, im:0f64};
    while is_in_range(zn) && depth < upper_iteration_limit() {
        zn = zn * zn + value;
        depth = depth + 1;
    }
    
    return depth;
}


pub fn is_in_range(value: Complex64) -> bool {
    let polar_form = value.to_polar();
    return polar_form.0.powf(2.0) <= upper_squared_limit() as f64;
}